### ISSUE TITLE
Reading an "ascii" table that is unicode

### ISSUE THREAD

**Issue by keflavich** [astropy/astropy#3826]
The ESO archive contains some tables that include unicode characters, e.g. csv tables.  For example, here's one in raw form:

```
'Release Date,Product ID,Instrument type,ProgId,PI/CoI,Proposal Title,APEX Partner Data,APEX Project ID\n2011-07-14,E-085.B-0964A.2010JUL08.TAR,Heterodyne,085.B-0964(A),HENKEL/ MAUERSBERGER/ M{\xc3\x83\xc2\x9c}HLE/ AALTO/ MANGUM/ MARTIN-PINTADO/ RIQUELME/ CUNNINGHAM/ JONES/ OTT/ REQUENA-TORRES/ BELLOCHE/ NAGAR/ MENTEN,The Kinetic Temperature of the Galactic Center Clouds,ESO data,E-085.B-0964A-2010\n2011-07-14,E-085.B-0964A.2010JUL08.TAR,Heterodyne,085.B-0964(A),HENKEL/ MAUERSBERGER/ MUHLE/ AALTO/ MANGUM/ MARTIN-PINTADO/ RIQUELME/ CUNNINGHAM/ JONES/ OTT/ REQUENA-TORRES/ BELLOCHE/ NAGAR/ MENTEN,The Kinetic Temperature of the Galactic Center Clouds,ESO data,E-085.B-0964A-2010\n\n# A total of 2 records were found matching the provided criteria.\n#'
```

It contains unicode for one author name.  I therefore get a parser error if I try to read this with the astropy ascii readers:

```
*** UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 24: ordinal not in range(128)
```

Any suggestions on how to deal with this?  It is what I think of as an "ascii table", but strictly it is not.  

cc @taldcroft, I figure you might have a suggestion?

**Comment by astrofrog** [comment]
In the past I know the attitude has been that io.ascii only deals with ASCII, but I agree that it might be time to start thinking about allowing unicode in some of the pure-Python readers at least.

**Comment by keflavich** [comment]
Probably `io.ascii` should be limited to ascii tables.  But I want to know if there is any way to ingest a (csv) table containing unicode characters into an astropy Table.  If not, it means that there are going to be tables in archives such as the ESO archive that are fundamentally incompatible with astropy.  

I'm guessing an answer is "find a way to strip out unicode characters, then treat as ascii", which I'm fine with if there's a reasonable way to do it.

**Comment by taldcroft** [comment]
In theory support for unicode tables should be possible in Python 3 because the csv module there works with unicode. That module does the heavy lifting in the pure Python io. Ascii.

**Comment by taldcroft** [comment]
@keflavich - your example works out of the box on Python 3.  So is using Python 3 an option?  Otherwise in Python 2 just take your unicode string and de-unicode it with `line.decode('utf-8').encode('ascii', errors='ignore')`
```
>>> line = 'Release Date,Product ID,Instrument type,ProgId,PI/CoI,Proposal Title,APEX Partner Data,APEX Project ID\n2011-07-14,E-085.B-0964A.2010JUL08.TAR,Heterodyne,085.B-0964(A),HENKEL/ MAUERSBERGER/ M{\xc3\x83\xc2\x9c}HLE/ AALTO/ MANGUM/ MARTIN-PINTADO/ RIQUELME/ CUNNINGHAM/ JONES/ OTT/ REQUENA-TORRES/ BELLOCHE/ NAGAR/ MENTEN,The Kinetic Temperature of the Galactic Center Clouds,ESO data,E-085.B-0964A-2010\n2011-07-14,E-085.B-0964A.2010JUL08.TAR,Heterodyne,085.B-0964(A),HENKEL/ MAUERSBERGER/ MUHLE/ AALTO/ MANGUM/ MARTIN-PINTADO/ RIQUELME/ CUNNINGHAM/ JONES/ OTT/ REQUENA-TORRES/ BELLOCHE/ NAGAR/ MENTEN,The Kinetic Temperature of the Galactic Center Clouds,ESO data,E-085.B-0964A-2010\n\n# A total of 2 records were found matching the provided criteria.\n#'

>>> ascii.read(line.decode('utf-8').encode('ascii', errors='replace'))
<Table masked=False length=2>
Release Date          Product ID         Instrument type ... APEX Partner Data  APEX Project ID  
  string80            string216              string80    ...      string64         string144     
------------ --------------------------- --------------- ... ----------------- ------------------
  2011-07-14 E-085.B-0964A.2010JUL08.TAR      Heterodyne ...          ESO data E-085.B-0964A-2010
  2011-07-14 E-085.B-0964A.2010JUL08.TAR      Heterodyne ...          ESO data E-085.B-0964A-2010
```

**Comment by keflavich** [comment]
@taldcroft - no, py3 isn't an option because this is part of astroquery, and I need astroquery to be py2.7 compatible.  

I'll try the workaround.  Thanks.

**Comment by embray** [comment]
I marked this as a "bug" in that in principle unicode should be handled better, but it's a pretty broad bug and isn't necessarily a 100% fixable issue.

**Comment by hamogu** [comment]
When are we going to drop 2.7 support?  ;-)

**Comment by taldcroft** [comment]
There was talk of allowing Py3-only features in astropy, in part to encourage use of Py3.  I would say that reading unicode tables counts as such a feature.  At some level that might be a reasonable argument for simply closing this issue as a `wontfix`.  @embray?

**Comment by astrofrog** [comment]
+1 to @taldcroft but I would at least add a notes to the docs to advertise that this works on Python 3

**Comment by hamogu** [comment]
:+1: (to closing as wontfix and adding and issue to the docs)

**Comment by keflavich** [comment]
If this is really unfixable in py2.7 - and I trust @taldcroft that it is - then marking it as such in the docs is acceptable.  It would be helpful to have an appropriate error message when reading ascii files in py2.7, e.g.: "A non-ascii character has been detected in the file.  Sometimes, text tables containing unicode can be read using astropy in python3; please see [this section of the docs] for details."

But please don't adopt the attitude of ignoring the needs of python2.7 users: there are plenty of tools out there that are permanently locked into python<3 and users do not have the option of upgrading.

**Comment by embray** [comment]
@keflavich I'm sure it's not *really* "unfixable".  I think it's just a matter of priorities and effort.  If someone wanted to try to fix it I'm sure they'd be welcome.

> But please don't adopt the attitude of ignoring the needs of python2.7 users: there are plenty of tools out there that are permanently locked into python<3 and users do not have the option of upgrading.

See #4454.

**Comment by embray** [comment]
For someone interested in implementing this feature on Python 2, the `csv` module documentation contains an example for implementing a unicode-based reader/writer for CSV:

https://docs.python.org/2/library/csv.html#examples

I've used this example myself in some projects.  I see no reason it couldn't be adapted in to `astropy.io.ascii` where needed.  Perhaps an "encoding" option could be added to the CSV reader.  If not given, or if "ascii" it could just go with the default, but otherwise it could use the `UnicodeReader`.

**Comment by taldcroft** [comment]
I had said `won't fix` not `can't fix`.  But I may have been hasty in saying that this issue should be closed, since as @embray notes it may not be terribly difficult.

**Comment by jpinedaf** [comment]
Hi, I'm trying to read into a CSV file from a Google Spreadsheet and I got the same issue. From the error message it was really hard to track it down to this issue. 

BTW, we are using python2.7.

**Comment by pllim** [comment]
I have the same error but in Python 3 using the latest dev. I am stumped. File attached (remove the `.txt` extension). Both `pandas` and `csv` can read the data successfully.

[Top5000population.csv.txt](https://github.com/astropy/astropy/files/551209/Top5000population.csv.txt)

```python
>>> cities = Table.read('Top5000population.csv', format='ascii.csv', guess=False, fast_reader=False)

.../astropy/table/table.py in read(cls, *args, **kwargs)
   2333         passed through to the underlying data reader (e.g. `~astropy.io.ascii.read`).
   2334         """
-> 2335         return io_registry.read(cls, *args, **kwargs)
   2336 
   2337     def write(self, *args, **kwargs):

.../astropy/io/registry.py in read(cls, *args, **kwargs)
    466 
    467         reader = get_reader(format, cls)
--> 468         data = reader(*args, **kwargs)
    469 
    470         if not isinstance(data, cls):

.../astropy/io/ascii/connect.py in io_read(format, filename, **kwargs)
     35     from .ui import read
     36     format = re.sub(r'^ascii\.', '', format)
---> 37     return read(filename, format=format, **kwargs)
     38 
     39 

.../astropy/io/ascii/ui.py in read(table, guess, **kwargs)
    339                                              ' with fast (no guessing)'})
    340         else:
--> 341             dat = reader.read(table)
    342             _read_trace.append({'kwargs': new_kwargs,
    343                                 'status': 'Success with specified Reader class '

.../astropy/io/ascii/core.py in read(self, table)
   1142 
   1143         # Get a list of the lines (rows) in the table
-> 1144         self.lines = self.inputter.get_lines(table)
   1145 
   1146         # Set self.data.data_lines to a slice of lines contain the data rows

.../astropy/io/ascii/core.py in get_lines(self, table)
    282                     ('\n' not in table + '' and '\r' not in table + '')):
    283                 with get_readable_fileobj(table) as file_obj:
--> 284                     table = file_obj.read()
    285             lines = table.splitlines()
    286         except TypeError:

.../python3.5/codecs.py in decode(self, input, final)
    319         # decode input (taking the buffer into account)
    320         data = self.buffer + input
--> 321         (result, consumed) = self._buffer_decode(data, self.errors, final)
    322         # keep undecoded input until the next call
    323         self.buffer = data[consumed:]

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa4 in position 39853: invalid start byte
```

**Comment by taldcroft** [comment]
@pllim - this issue is unrelated to io.ascii.  That file is not utf-8 encoded.  I tried latin1 and that also fails to reproduce the expected ñ character.
```
In [28]: lines = open(fn, 'rb').read()

In [29]: lines[39848:39860].decode('latin1')
Out[29]: 'La Ca¤ada Fl'

In [30]: lines[39848:39860].decode('utf-8')
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-30-e54bdb8d6b7e> in <module>()
----> 1 lines[39848:39860].decode('utf-8')

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa4 in position 5: invalid start byte

In [31]: lines = open(fn, 'r').read()
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-40-aa2a38461205> in <module>()
----> 1 lines = open(fn, 'r').read()

/Users/aldcroft/anaconda/envs/astropy3/lib/python3.5/codecs.py in decode(self, input, final)
    319         # decode input (taking the buffer into account)
    320         data = self.buffer + input
--> 321         (result, consumed) = self._buffer_decode(data, self.errors, final)
    322         # keep undecoded input until the next call
    323         self.buffer = data[consumed:]

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa4 in position 39853: invalid start byte

```

**Comment by pllim** [comment]
@taldcroft , thanks for the quick reply! Is there a workaround for reading this file with Astropy table at all? Just wondering. It was given out as a sample data in a Python class. I can change that bad row easy enough but everyone who has that file will run into the same problem with Astropy.

**Comment by taldcroft** [comment]
@pllim 
```
In [43]: text = open(fn, 'rb').read().decode('ascii', errors='replace')
In [44]: dat = ascii.read(text)
```

**Comment by saimn** [comment]
It seems to be latin1:
```
In [16]: !file Top5000population.csv.txt
Top5000population.csv.txt: ISO-8859 text, with CRLF line terminators
```
It would be useful to have the possibility to specify an encoding to `ascii.read`/`Table.read`. Here I added this possibility to `Table.read` when `guess` is True, which is the easiest part (then it should also be passed to the readers' `read` method):
```python
In [18]: Table.read('Top5000population.csv.txt', format='csv', fast_reader=False, encoding='lat
    ...: in1')[1648:]
Out[18]:
<Table length=3352>
        city         state population
       str30          str2    str9
-------------------- ----- ----------
La Ca¤ada Flintridge    CA     20,671
       Junction City    KS     20,671
        West Mifflin    PA     20,669
            Lockport    NY     20,630
              Oxford    AL     20,622
         Lake Zurich    IL     20,610
           Riverbank    CA     20,606
...
```
Patch:
```diff
diff --git a/astropy/io/ascii/ui.py b/astropy/io/ascii/ui.py
index 41459d2..f554e1a 100644
--- a/astropy/io/ascii/ui.py
+++ b/astropy/io/ascii/ui.py
@@ -180,7 +180,7 @@ def _get_format_class(format, ReaderWriter, label):
     return ReaderWriter


-def read(table, guess=None, **kwargs):
+def read(table, guess=None, encoding=None, **kwargs):
     """
     Read the input ``table`` and return the table.  Most of
     the default behavior for various parameters is determined by the Reader
@@ -285,7 +285,7 @@ def read(table, guess=None, **kwargs):
         # intact.
         if 'readme' not in new_kwargs:
             try:
-                with get_readable_fileobj(table) as fileobj:
+                with get_readable_fileobj(table, encoding=encoding) as fileobj:
                     table = fileobj.read()
             except ValueError:  # unreadable or invalid binary file
                 raise
```

**Comment by taldcroft** [comment]
@saimn - off topic from the real issue, but I believe the city in question is La Cañada, not La Ca¤ada.  So either the file is corrupted or it is not Latin1.  😝

**Comment by saimn** [comment]
@taldcroft - Yep, but at least the file is readable :wink:

**Comment by taldcroft** [comment]
@saimn - PR in work??   :smile:

**Comment by pllim** [comment]
:+1: to PR. That would be nicer than having to manually fix the file...

**Comment by saimn** [comment]
It is in the queue, but it will need some work on how to pass the encoding to the different places where the file is opened (the current patch works only when guess is True).

**Comment by saimn** [comment]
Can be closed ? #5448 add the encoding parameter for Python 3, but it will probably never be fixed for Python 2 (unless someone really wants to !).

**Comment by hamogu** [comment]
👍  to closing. The 2.0 release coming up will be the last one to support python 2, so after that there is no need to fix this for python 2 any longer.
